src/plugins/intel_cpu/src/nodes/segment_max.cpp

```diff
@@ -0,0 +1,191 @@
+    if (getOriginalInputsNumber() == 2) {
+        addSupportedPrimDesc({{LayoutType::ncsp, dataPrecision}, {LayoutType::ncsp, segmentIdsPrecision}},
+                             {{LayoutType::ncsp, dataPrecision}},
+                             impl_desc_type::ref);
+    } else {
+        ov::element::Type numSegmentsPrecision = getOriginalInputPrecisionAtPort(2);
+        addSupportedPrimDesc({{LayoutType::ncsp, dataPrecision},
+                              {LayoutType::ncsp, segmentIdsPrecision},
+                              {LayoutType::ncsp, numSegmentsPrecision}},
+                             {{LayoutType::ncsp, dataPrecision}},
+                             impl_desc_type::ref);
+    }
+}
+
+bool SegmentMax::created() const {
+    return getType() == Type::SegmentMax;
+}
+
+bool SegmentMax::needPrepareParams() const {
+    return false;
+}
+
+void SegmentMax::executeDynamicImpl(const dnnl::stream& strm) {
+    execute(strm);
+
+    // Update lastSegmentIds
+    std::vector<int64_t> segmentIds(getSrcMemoryAtPort(1)->getSize());
+    auto segmentIdsPrecision = getOriginalInputPrecisionAtPort(1);
+    if (segmentIdsPrecision == ov::element::i32) {
+        const int32_t* srcSegmentIds = getSrcDataAtPortAs<const int32_t>(1);
+        std::copy(srcSegmentIds, srcSegmentIds + segmentIds.size(), segmentIds.begin());
+    } else if (segmentIdsPrecision == ov::element::i64) {
+        const int64_t* srcSegmentIds = getSrcDataAtPortAs<const int64_t>(1);
+        std::copy(srcSegmentIds, srcSegmentIds + segmentIds.size(), segmentIds.begin());
+    } else {
+        OPENVINO_THROW("Unsupported index type for segment_ids input");
+    }
+    lastSegmentIds.assign(segmentIds.begin(), segmentIds.end());
+
+    // Update lastNumSegments
+    if (getOriginalInputsNumber() == 3) {
+        const auto& numSegmentsPrecision = getOriginalInputPrecisionAtPort(2);
+        int64_t numSegmentsValue;
+        if (numSegmentsPrecision == ov::element::i64) {
+            numSegmentsValue = *getSrcDataAtPortAs<const int64_t>(2);
+        } else if (numSegmentsPrecision == ov::element::i32) {
+            numSegmentsValue = static_cast<int64_t>(*getSrcDataAtPortAs<const int32_t>(2));
+        } else {
+            OPENVINO_THROW("Unsupported index type for num_segments input");
+        }
+        if (lastNumSegments.empty()) {
+            lastNumSegments.push_back(numSegmentsValue);
+        } else {
+            lastNumSegments[0] = numSegmentsValue;
+        }
+    }
+}
+
+bool SegmentMax::needShapeInfer() const {
+    if (inputShapesModified()) {
+        return true;
+    }
+    if (lastSegmentIds.empty()) {
+        return true;
+    }
+
+    // Check if segmentIds has changed
+    auto segmentIdsPrecision = getOriginalInputPrecisionAtPort(1);
+    const int64_t* segmentIds = nullptr;
+    if (segmentIdsPrecision == ov::element::i32) {
+        segmentIds = reinterpret_cast<const int64_t*>(getSrcDataAtPortAs<const int32_t>(1));
```

Comment:
It's not safe to `reinterpret_cast` i32 ptr to i64 ptr. This is what I underlined in the previous comments. 
It can be done for values, but is really risky for the pointers.
Dereferencing the int64_t* pointer will attempt to read 8 bytes from memory, even though only 4 bytes were allocated for each int32_t. This can lead to undefined behavior or incorrect results.

Reply:
Asserted that it's always i32 and corrected.